components/cop_codegen/src/rpn_function.rs

```diff
@@ -1,49 +1,319 @@
-    meta: Vec<Ident>,
+use syn::punctuated::Punctuated;
+use syn::*;
+
+/// Parses an attribute like `#[rpn_fn(vargs, capture = [ctx, output_rows])`.
+#[derive(Debug, Default)]
+struct RpnFnAttr {
+    is_varg: bool,
+    captures: Vec<Expr>,
+}
+
+impl parse::Parse for RpnFnAttr {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        let mut is_varg = false;
+        let mut captures = Vec::new();
+
+        let config_items = Punctuated::<Expr, Token![,]>::parse_terminated(input).unwrap();
+        for item in config_items {
+            match item {
+                Expr::Assign(ExprAssign { left, right, .. }) => {
+                    let left_str = format!("{}", (&left).into_token_stream());
+                    match left_str.as_ref() {
+                        "capture" => {
+                            if let Expr::Array(ExprArray { elems, .. }) = *right {
+                                captures = elems.into_iter().collect();
+                            } else {
+                                return Err(Error::new_spanned(right, "Expect array expression"));
+                            }
+                        }
+                        _ => {
+                            return Err(Error::new_spanned(
+                                left,
+                                format!("Unknown attribute parameter `{}`", left_str),
+                            ));
+                        }
+                    }
+                }
+                Expr::Path(ExprPath { path, .. }) => {
+                    let path_str = format!("{}", (&path).into_token_stream());
+                    match path_str.as_ref() {
+                        "varg" => {
+                            is_varg = true;
+                        }
+                        _ => {
+                            return Err(Error::new_spanned(
+                                path,
+                                format!("Unknown attribute parameter `{}`", path_str),
+                            ));
+                        }
+                    }
+                }
+                _ => {
+                    return Err(Error::new_spanned(
+                        item,
+                        "Expect attributes to be `foo=bar` or `foo`",
+                    ));
+                }
+            }
+        }
+        Ok(Self { is_varg, captures })
+    }
+}
+
+mod kw {
+    syn::custom_keyword!(Option);
+}
+
+/// Parses an evaluable type like `Option<T>`.
+struct RpnFnEvaluableType {
+    eval_type: TypePath,
+}
+
+impl parse::Parse for RpnFnEvaluableType {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        input.parse::<self::kw::Option>()?;
+        input.parse::<Token![<]>()?;
+        let eval_type = input.parse::<TypePath>()?;
+        input.parse::<Token![>]>()?;
+        Ok(Self { eval_type })
+    }
+}
+
+/// Parses a function signature parameter like `val: &Option<T>`.
+struct RpnFnSignatureParam {
+    _pat: Pat,
```

Comment:
Why not remove this field?

Reply:
Maybe useful in future, removing it looks to be wired because it accepts much info but only stores part of it.